README.md
# MOD_SimplePortal
ФОРК библиотеки "SimplePortal". Простой WiFi менеджер с возможностью задания Логина/Пароля для WEB интерфейса

### Совместимость
esp8266, esp32

## Содержание
- [Инициализация](#init)
- [Использование](#usage)
- [Пример](#example)
- [История изменений](#history)


<a id="init"></a>
## Инициализация
Нет

<a id="usage"></a>
## Использование
```cpp
void portalStart();     // запустить портал
void portalStop();      // остановить портал
bool portalTick();      // вызывать в цикле
void portalRun(uint32_t prd = 60000);   // блокирующий вызов
byte portalStatus();    // статус: 1 connect, 2 ap, 3 local, 4 exit, 5 timeout

// константы статуса
SP_ERROR            // ошибка
SP_SUBMIT           // отправлены логин-пароль
SP_SWITCH_AP        // сигнал на смену в AP
SP_SWITCH_LOCAL     // сигнал на смену в Local
SP_EXIT             // нажата кнопка выход
SP_TIMEOUT          // вышел таймаут
```

### Хранение настроек
Из скетча доступна переменная (структура) **portalCfg**:
- `char* SSID[32]` - логин
- `char* pass[32]` - пароль
- `byte mode` - (1 `WIFI_STA`, 2 `WIFI_AP`)
После отправки данных с портала можно забрать информацию из структуры и/или записать её в EEPROM.

### Как работает
![demo](/doc/demo.png)
- Библиотека запускает режим **SoftAP**, запускает **DNSServer** и **ESP8266WebServer**.
- Коннектимся к точке, по умолчанию называется **"ESP Config"**
- Откроется страница конфигурации (если не открылась - перейди по *192.168.4.1*). На странице можно 
настроить режим работы ESP, а также сконфигурировать подключение к роутеру (логин-пароль).
- Кнопка **"Сохранить ✓"** отправляет введённые данные на ESP. После нажатия **"Сохранить ✓"** режим в конфиге переключается в 1 (WIFI_STA).
- Библиотека не переключает итоговый режим работы esp, кнопки портала **"Режим AP"** / **"Режим STA"** просто дают сигналы "в скетч".
- Кнопка **"Закрыть портал"** просто завершает работу портала.
- После нажатия любой кнопки или выхода таймаута портал закрывается, сервер останавливается, SoftAP отключается.

<a id="example"></a>
## Примеры
### Блокирующий
```cpp
#include <ESP8266WiFi.h>
#include <SimplePortal.h>

void setup() {
  Serial.begin(9600);
  delay(3000);
  
  portalRun();  // запустить с таймаутом 60с
  //portalRun(30000); // запустить с кастомным таймаутом
  
  Serial.println(portalStatus());
  // статус: 0 error, 1 connect, 2 ap, 3 local, 4 exit, 5 timeout
  
  if (portalStatus() == SP_SUBMIT) {
    Serial.println(portalCfg.SSID);
    Serial.println(portalCfg.pass);
    // забираем логин-пароль
  }
}

void loop() {
}
```

### Асинхронный
```cpp
#include <ESP8266WiFi.h>
#include <SimplePortal.h>

void setup() {
  Serial.begin(9600);

  // запускаем портал
  portalStart();
}

void loop() {
  // вызываем в loop
  if (portalTick()) {
    Serial.println(portalStatus());
    if (portalStatus() == SP_SUBMIT) {
      Serial.println(portalCfg.SSID);
      Serial.println(portalCfg.pass);
      // забираем логин-пароль
    }

    // сработает однократно при действии
    // точка будет автоматически выключена
  }
}
```


### Получение данных WiFi и Логина/Пароля (Блокирующий)
```cpp
#include <ESP8266WiFi.h>
#include <MOD_SimplePortal.h>
#include <EEPROM.h>


void setup() {
  Serial.begin(115200);  // Инициализация Serial

  portalRun(300000);  // Pапустить с таймаутом 5 минут


  if (portalStatus() == SP_SUBMIT) {  // Если кнопка "Сохранить" с портала нажата, то получаем данные с формы через структуру
    EEPROM.begin(120);                // Инициализируем EEPROM
    EEPROM.put(0, portalCfg);         // Подготавливаем структуру к записи с нулевой ячейки (сохраняем её в ОЗУ)
    EEPROM.commit();                  // Записываем из ОЗУ в EEPROM
    EEPROM.end();                     // И очищаем данные из ОЗУ
    WiFi.softAPdisconnect();          // Отключаем режим AP
  }



  //Читаем логин пароль из памяти
  EEPROM.begin(120);         // Инициализируем EEPROM
  EEPROM.get(0, portalCfg);  // Считываем структуру начиная с нулевой ячейки


  Serial.print("SSID: ");
  Serial.println(portalCfg.ssid);
  Serial.print("Pass WiFi: ");
  Serial.println(portalCfg.pass);
  Serial.print("Login UI: ");
  Serial.println(portalCfg.loginUI);
  Serial.print("Pass UI: ");
  Serial.println(portalCfg.passUI);


  //Подключаемся к WiFi
  Serial.print("Connect to: ");
  Serial.println(portalCfg.ssid);
  WiFi.mode(WIFI_STA);                         // Запускаем WiFi в режиме "Станция"
  WiFi.hostname("ESP-TEST");                   // Задаём имя хоста
  WiFi.begin(portalCfg.ssid, portalCfg.pass);  // Берём данные из структуры
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }
  Serial.println();
  Serial.print("Connected! Local IP: ");
  Serial.println(WiFi.localIP());
}


void loop() {
}

```

<a id="history"></a>
## История изменений находится в **"CHANGELOG"**

  Дата изменения: **01.03.2023г.**
  
  Версия: **0.1**
  
  Автор: **Otto**


CHANGELOG
*  ФОРК библиотеки "SimplePortal". Простой WiFi менеджер с возможностью задания Логина/Пароля для WEB интерфейса


MOD_SimplePortal 0.1 - 01.03.2022

* Добавлены поля Логин/Пароль (для UI или других целей) на WEB интерфейсе.
Получение данных из структуры такое же как в оригинале, только добавились ещё 2 поля с логином и паролем от UI.
* Немного украсил WEB страничку.
* Русифицировал и исправил кодировку (что бы корректно отображалась Кирилица).



src/MOD_SimplePortal.cpp
#include "MOD_SimplePortal.h"
static DNSServer _SP_dnsServer;
#ifdef ESP8266
static ESP8266WebServer _SP_server(80);
#else
static WebServer _SP_server(80);
#endif

//WEB интерфейс
const char SP_connect_page[] PROGMEM = R"rawliteral(
<!DOCTYPE HTML>
<html><head>
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta charset="UTF-8" />
</head><body>
<style type="text/css">
	body { background-color:#848483; }
    input[type="text"] { margin-bottom:8px;font-size:20px; }
    input[type="submit"] { width:180px; height:60px;margin-bottom:8px;font-size:20px; }
    .BtSave { background: #32CD32; }
    .BtAcPo { background: #F7941E; }
    .BtLoMo { background: #008EB0; }
    .BtExPr { background: #C1172C; }
</style>
<center>
<h3>Настройки WiFi</h3>
<form action="/connect" method="POST">
    <input type="text" name="ssid" placeholder="Введите SSID"><br>
    <input type="text" name="pass" placeholder="Введите Пароль">
<h3>Настройки Авторизации</h3>
	<input type="text" name="loginUI" placeholder="Введите Логин"><br>
	<input type="text" name="passUI" placeholder="Введите Пароль">
    <br>
    <input type="submit" class="BtSave" value="Сохранить ✓">
</form>
<h3>Переключить режим WiFi</h3>
<form action="/ap" method="POST">
    <input type="submit" class="BtAcPo" value="Режим AP">
</form>
<form action="/local" method="POST">
    <input type="submit" class="BtLoMo" value="Режим STA">
</form>
<form action="/exit" method="POST">
    <input type="submit" class="BtExPr" value="Закрыть портал">
</form>
</center>
</body></html>)rawliteral";


static bool _SP_started = false;
static byte _SP_status = 0;
PortalCfg portalCfg;

void SP_handleConnect() {
  strcpy(portalCfg.ssid, _SP_server.arg("ssid").c_str());
  strcpy(portalCfg.pass, _SP_server.arg("pass").c_str());
  strcpy(portalCfg.loginUI, _SP_server.arg("loginUI").c_str());
  strcpy(portalCfg.passUI, _SP_server.arg("passUI").c_str());
  portalCfg.mode = WIFI_STA;
  _SP_status = 1;
}

void SP_handleAP() {
  portalCfg.mode = WIFI_AP;
  _SP_status = 2;
}

void SP_handleLocal() {
  portalCfg.mode = WIFI_STA;
  _SP_status = 3;
}

void SP_handleExit() {
  _SP_status = 4;
}

void portalStart() {
  WiFi.softAPdisconnect();
  WiFi.disconnect();
  IPAddress apIP(SP_AP_IP);
  IPAddress subnet(255, 255, 255, 0);
  WiFi.mode(WIFI_AP);
  WiFi.softAPConfig(apIP, apIP, subnet);
  WiFi.softAP(SP_AP_NAME);
  _SP_dnsServer.start(53, "*", apIP);

  _SP_server.onNotFound([]() {
    _SP_server.send(200, "text/html", SP_connect_page);
  });
  _SP_server.on("/connect", HTTP_POST, SP_handleConnect);
  _SP_server.on("/ap", HTTP_POST, SP_handleAP);
  _SP_server.on("/local", HTTP_POST, SP_handleLocal);
  _SP_server.on("/exit", HTTP_POST, SP_handleExit);
  _SP_server.begin();
  _SP_started = true;
  _SP_status = 0;
}

void portalStop() {
  WiFi.softAPdisconnect();
  _SP_server.stop();
  _SP_dnsServer.stop();
  _SP_started = false;
}

bool portalTick() {
  if (_SP_started) {
    _SP_dnsServer.processNextRequest();
    _SP_server.handleClient();
    yield();
    if (_SP_status) {
      portalStop();
      return 1;
    }
  }
  return 0;
}

void portalRun(uint32_t prd) {
  uint32_t tmr = millis();
  portalStart();
  while (!portalTick()) {
    if (millis() - tmr > prd) {
      _SP_status = 5;
      portalStop();
      break;
    }
    yield();
  }
}

byte portalStatus() {
  return _SP_status;
}

src/MOD_SimplePortal.h
/*
    ФОРК библиотеки "SimplePortal".
	Простой WiFi менеджер с возможностью задания Логина/Пароля для WEB интерфейса
		
    GitHub: https://github.com/Otto17/MOD_SimplePortal
    

    Версия:
    ver 0.1

*/

#define SP_AP_NAME "ESP Config"		// Название точки доступа
#define SP_AP_IP 192,168,4,1        	// IP точки доступа

#ifndef _MOD_SimplePortal_h
#define _MOD_SimplePortal_h
#include <DNSServer.h>
#ifdef ESP8266
#include <ESP8266WiFi.h>
#include <ESP8266WebServer.h>
#else
#include <WiFi.h>
#include <WebServer.h>
#endif

#define SP_ERROR 0
#define SP_SUBMIT 1
#define SP_SWITCH_AP 2
#define SP_SWITCH_LOCAL 3
#define SP_EXIT 4
#define SP_TIMEOUT 5

//Структура для получения с WEB интерфейса данных
struct PortalCfg {
  char ssid[32];
  char pass[32];
  char loginUI[20];
  char passUI[20];
  uint8_t mode = WIFI_AP;    // (1 WIFI_STA, 2 WIFI_AP)
};
extern PortalCfg portalCfg;

void portalStart();     // Запустить портал
void portalStop();      // Остановить портал
bool portalTick();      // Вызывать в цикле
void portalRun(uint32_t prd = 60000);   // Блокирующий вызов
byte portalStatus();    // Статус: 1 connect, 2 ap, 3 local, 4 exit, 5 timeout

void SP_handleConnect();
void SP_handleAP();
void SP_handleLocal();
void SP_handleExit();
#endif
